File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter09/AnimatedPieSlice/AnimatedPieSlice/PieSlice.cs
using System;
using Windows.Foundation;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Media;
using Windows.UI.Xaml.Shapes;

namespace AnimatedPieSlice
{
    public class PieSlice : Path
    {
        PathFigure pathFigure;
        LineSegment lineSegment;
        ArcSegment arcSegment;

        static PieSlice()
        {
            CenterProperty = DependencyProperty.Register("Center",
                typeof(Point), typeof(PieSlice),
                new PropertyMetadata(new Point(100, 100), OnPropertyChanged));

            RadiusProperty = DependencyProperty.Register("Radius",
                typeof(double), typeof(PieSlice),
                new PropertyMetadata(100.0, OnPropertyChanged));

            StartAngleProperty = DependencyProperty.Register("StartAngle",
                typeof(double), typeof(PieSlice),
                new PropertyMetadata(0.0, OnPropertyChanged));

            SweepAngleProperty = DependencyProperty.Register("SweepAngle",
                typeof(double), typeof(PieSlice),
                new PropertyMetadata(90.0, OnPropertyChanged));
        }

        public PieSlice()
        {
            pathFigure = new PathFigure { IsClosed = true };
            lineSegment = new LineSegment();
            arcSegment = new ArcSegment { SweepDirection = SweepDirection.Clockwise };
            pathFigure.Segments.Add(lineSegment);
            pathFigure.Segments.Add(arcSegment);

            PathGeometry pathGeometry = new PathGeometry();
            pathGeometry.Figures.Add(pathFigure);

            this.Data = pathGeometry;
            UpdateValues();
        }

        public static DependencyProperty CenterProperty { private set; get; }

        public static DependencyProperty RadiusProperty { private set; get; }

        public static DependencyProperty StartAngleProperty { private set; get; }

        public static DependencyProperty SweepAngleProperty { private set; get; }

        public Point Center
        {
            set { SetValue(CenterProperty, value); }
            get { return (Point)GetValue(CenterProperty); }
        }

        public double Radius
        {
            set { SetValue(RadiusProperty, value); }
            get { return (double)GetValue(RadiusProperty); }
        }

        public double StartAngle
        {
            set { SetValue(StartAngleProperty, value); }
            get { return (double)GetValue(StartAngleProperty); }
        }

        public double SweepAngle
        {
            set { SetValue(SweepAngleProperty, value); }
            get { return (double)GetValue(SweepAngleProperty); }
        }

        static void OnPropertyChanged(DependencyObject obj, 
                                      DependencyPropertyChangedEventArgs args)
        {
            (obj as PieSlice).UpdateValues();
        }

        void UpdateValues()
        {
            pathFigure.StartPoint = this.Center;

            double x = this.Center.X + this.Radius * Math.Sin(Math.PI * this.StartAngle / 180);
            double y = this.Center.Y - this.Radius * Math.Cos(Math.PI * this.StartAngle / 180);
            lineSegment.Point = new Point(x, y);

            x = this.Center.X + this.Radius * Math.Sin(Math.PI * (this.StartAngle + 
                                                                  this.SweepAngle) / 180);

            y = this.Center.Y - this.Radius * Math.Cos(Math.PI * (this.StartAngle + 
                                                                  this.SweepAngle) / 180);
            arcSegment.Point = new Point(x, y);
            arcSegment.IsLargeArc = this.SweepAngle >= 180;

            arcSegment.Size = new Size(this.Radius, this.Radius);
        }
    }
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter09/AnimatedPieSlice/AnimatedPieSlice/PieSlice.cpp
#include "pch.h"
#include "PieSlice.h"

using namespace AnimatedPieSlice;

using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::System;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Interop;
using namespace Windows::UI::Xaml::Media;

DependencyProperty^ PieSlice::centerProperty = 
    DependencyProperty::Register("Center", 
                                 TypeName(Point::typeid), 
                                 TypeName(PieSlice::typeid),
                                 ref new PropertyMetadata(Point(100, 100), 
                                 ref new PropertyChangedCallback(OnPropertyChanged)));

DependencyProperty^ PieSlice::radiusProperty = 
    DependencyProperty::Register("Radius", 
                                 TypeName(double::typeid), 
                                 TypeName(PieSlice::typeid),
                                 ref new PropertyMetadata(100.0, 
                                 ref new PropertyChangedCallback(OnPropertyChanged)));

DependencyProperty^ PieSlice::startAngleProperty = 
    DependencyProperty::Register("StartAngle", 
                                 TypeName(double::typeid), 
                                 TypeName(PieSlice::typeid),
                                 ref new PropertyMetadata(0.0, 
                                 ref new PropertyChangedCallback(OnPropertyChanged)));

DependencyProperty^ PieSlice::sweepAngleProperty = 
    DependencyProperty::Register("SweepAngle", 
                                 TypeName(double::typeid), 
                                 TypeName(PieSlice::typeid),
                                 ref new PropertyMetadata(90.0, 
                                 ref new PropertyChangedCallback(OnPropertyChanged)));

PieSlice::PieSlice()
{
    lineSegment = ref new LineSegment();
    arcSegment = ref new ArcSegment();
    arcSegment->SweepDirection = SweepDirection::Clockwise;

    pathFigure = ref new PathFigure();
    pathFigure->IsClosed = true;
    pathFigure->Segments->Append(lineSegment);
    pathFigure->Segments->Append(arcSegment);

    PathGeometry^ pathGeometry = ref new PathGeometry();
    pathGeometry->Figures->Append(pathFigure);

    this->Data = pathGeometry;
    UpdateValues();
}

DependencyProperty^ PieSlice::CenterProperty::get()
{
    return centerProperty;
}

DependencyProperty^ PieSlice::RadiusProperty::get()
{
    return radiusProperty;
}

DependencyProperty^ PieSlice::StartAngleProperty::get()
{
    return startAngleProperty;
}

DependencyProperty^ PieSlice::SweepAngleProperty::get()
{
    return sweepAngleProperty;
}

void PieSlice::Center::set(Point point)
{
    SetValue(CenterProperty, point);
}

Point PieSlice::Center::get()
{
    return (Point)GetValue(CenterProperty); 
}

void PieSlice::Radius::set(double radius)
{
    SetValue(RadiusProperty, radius);
}

double PieSlice::Radius::get()
{
    return (double)GetValue(RadiusProperty);
}

void PieSlice::StartAngle::set(double angle)
{
    SetValue(StartAngleProperty, angle);
}

double PieSlice::StartAngle::get()
{
    return (double)GetValue(StartAngleProperty);
}

void PieSlice::SweepAngle::set(double angle)
{
    SetValue(SweepAngleProperty, angle);
}

double PieSlice::SweepAngle::get()
{
    return (double)GetValue(SweepAngleProperty);
}

void PieSlice::OnPropertyChanged(DependencyObject^ obj,
                                 DependencyPropertyChangedEventArgs^ args)
{
    dynamic_cast<PieSlice^>(obj)->UpdateValues();
}

void PieSlice::UpdateValues()
{
    pathFigure->StartPoint = this->Center;

    double radians = 3.14159 * this->StartAngle / 180;
    double x = this->Center.X + this->Radius * sin(radians);
    double y = this->Center.Y - this->Radius * cos(radians);
    lineSegment->Point = Point((float)x, (float)y);

    radians += 3.14159 * this->SweepAngle / 180;
    x = this->Center.X + this->Radius * sin(radians);
    y = this->Center.Y - this->Radius * cos(radians);

    arcSegment->Point = Point((float)x, (float)y);
    arcSegment->IsLargeArc = this->SweepAngle >= 180;
    arcSegment->Size = Size((float)this->Radius, (float)this->Radius);
}
